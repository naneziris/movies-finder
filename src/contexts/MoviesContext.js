import React, { useState, useEffect, useContext, useCallback } from 'react';
import _isEmpty from 'lodash/isEmpty';
import { MOVIES_IN_THEATER_URL, MOVIES_SEARCH_URL, MOVIES_GENRES_URL } from '../config/constants';

const MoviesContext = React.createContext();

const MoviesProvider = (props) => {

    const [movies, setMovies] = useState([]);
    const [page, setPage] = useState(1);
    const [totalPages, setTotalPages] = useState([]);
    const [listName, setListName] = useState(['in-theaters']);
    const [genres, setGenres] = useState({});
    const [loading, setLoading] = useState(true);
    const [showInTheatersButton, setShowInTheatersButton] = useState(false);
    const [search, setSearch] = useState('');

    /**
     * Fetches a list of all genres
     * Flattens the object and sets it at the Context
     */
    const fetchGenres = useCallback(async () => {
        // Genres do not change often, so If we have already fetched them and placed in Context
        // do not fetch again
        if (!_isEmpty(genres)) return;
        try {
            const genresData = await fetch(MOVIES_GENRES_URL);
            const genres = await genresData.json();
            const flattenedGenresObject = Object.assign({}, ...genres.genres.map(object => ({[object.id]: object.name})));
            setGenres(flattenedGenresObject);
            setLoading(false);
        } catch (e) {
            console.log(e);
        }
    }, [genres]);

    /**
     * Fetches the movies that are "in theater now"
     * Returns array objects
     * @param {number} page 
     */
    const fetchMovies = useCallback( async (page = 1) => {
        try {
            const movieData = await fetch(`${MOVIES_IN_THEATER_URL}&page=${page}`);
            const moviesInTheater = await movieData.json();
            setMovies(moviesInTheater.results);
            setPage(moviesInTheater.page);
            setTotalPages(moviesInTheater.total_pages);
            setListName('in-theaters');
            setLoading(false);
        } catch (e) {
            console.log(e);
        }
    }, []);


    /**
     * Debounce method used for limiting the number of
     * requests made from the handleSearch
     * @param {function} fn 
     * @param {*number} time 
     */
    const debounce = (fn, time) => {
        let timeout;

        return function() {
            const functionCall = () => fn.apply(this, arguments);

            clearTimeout(timeout);
            timeout = setTimeout(functionCall, time);
        };
    };

    /**
     * Returns an array of movies objects based on the search term (value)
     * and the requested page
     * @param {string} value 
     * @param {*number} page 
     */
    const handleSearch = useCallback( async (value, page = 1) => {
        if (!value || value === '') {
            setSearch(null);
            return fetchMovies();
        }
        try {
            setLoading(true);
            const response = await fetch(`${MOVIES_SEARCH_URL}&query=${value}&page=${page}`);
            const searchRelatedMovies = await response.json();
            setMovies(searchRelatedMovies.results);
            setPage(searchRelatedMovies.page);
            setTotalPages(searchRelatedMovies.total_pages);
            setListName('search-results');
            setSearch(value);
            setLoading(false);
            setShowInTheatersButton(true);
        } catch (e) {
            console.log(e);
        }
    }, []);

    /**
     * Fetches the next page of the movies list
     * If current listName is 'search-results' it uses the "movie search" endpoint
     * else it uses the "in theater now" endpoint
     * Pushes the new movie objects to the existing array of objects in Context
     * @param {number} page 
     */
    const fetchMoreMovies = useCallback( async (page = 2) => {
        let movieData;
        try {
            if(listName === 'search-results') {
                movieData = await fetch(`${MOVIES_SEARCH_URL}&query=${search}&page=${page}`);
            } else {
                movieData = await fetch(`${MOVIES_IN_THEATER_URL}&page=${page}`);
            }
            const newMovies = await movieData.json();
            setMovies([...movies, ...newMovies.results]);
            setPage(newMovies.page);
            setTotalPages(newMovies.total_pages);
            setLoading(false);
        } catch (e) {
            console.log(e);
        }
    }, [listName, movies, search]);

    const debounceHandleSearch = useCallback(debounce(handleSearch, 900), []);

    useEffect(() => {
        fetchMovies();
        fetchGenres();
    }, []);
    
      return (
        <MoviesContext.Provider value={{
            loading,
            search,
            movies,
            genres,
            page,
            listName,
            totalPages,
            fetchGenres,
            fetchMovies,
            debounceHandleSearch,
            showInTheatersButton,
            setShowInTheatersButton,
            setSearch,
            setPage,
            setTotalPages,
            fetchMoreMovies,
        }}>
            {props.children}
        </MoviesContext.Provider>
      )
    }
/**
 * Custom hook to consume the MoviesContext
 */
function useMoviesContext() {
    return useContext(MoviesContext);
}

export { MoviesProvider,  useMoviesContext }