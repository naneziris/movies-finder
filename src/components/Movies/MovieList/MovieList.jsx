import React, { memo } from 'react';
import InfiniteScroll from "react-infinite-scroller";
import _isEqual from 'lodash/isEqual';
import MovieItem from '../MovieItem/MovieItem';
import { useMoviesContext } from '../../../contexts/MoviesContext';
import '../Movies.css';

export const MovieList = () => {
    const { movies, loading, page, totalPages, fetchMoreMovies, genres } = useMoviesContext();
    const emptyResults = movies?.length === 0;

    const showItemGenres = genresIdsArray => (
        <>
            {
                genresIdsArray.map(id => {
                    return <span className="movies__genre" key={id}>{genres[id] || ''}</span>;
                })
            }
        </>
    )

    return (
        <>
            <div>
                {loading ?
                    <h4 className="text-center">
                        ...loading
                    </h4> :
                    <ul className="movies-container">
                        <InfiniteScroll
                            pageStart={0}
                            loadMore={() => fetchMoreMovies(page + 1)}
                            hasMore={totalPages > page}
                            loader={
                                <div className="loader" key={0}>
                                Loading ...
                                </div>
                            }
                        >
                            {movies.map((movie, i) => {
                                return <MovieItem key={`${movie.id}-${i}`} movie={movie} genres={showItemGenres(movie['genre_ids'])} />
                            })}
                        </InfiniteScroll>
                    </ul>
                }
                {emptyResults &&
                    <h4 className="movies__empty">Sorry, there are no results</h4>
                }
            </div>
        </>
    );
};

const areEqual = (prevProps, nextProps) => {
    return (
        _isEqual(prevProps.movies, nextProps.movies) &&
        _isEqual(prevProps.genres, nextProps.genres) &&
        prevProps.fetchMoreMovies === nextProps.fetchMoreMovies &&
        prevProps.loading === nextProps.loading &&
        prevProps.page === nextProps.page
    );
};

export default memo(MovieList, areEqual);