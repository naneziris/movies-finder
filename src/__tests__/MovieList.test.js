import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { mount, configure, shallow } from 'enzyme';
import MovieList from '../components/Movies/MovieList/MovieList';
import * as MoviesContext from '../contexts/MoviesContext';

configure({adapter: new Adapter()});

jest.mock('../contexts/MoviesContext', () => ({useMoviesContext: jest.fn()}));

describe('MovieList', () => {
test('Should display an h4 loading text if loading is set to true', () => {
    const contextValues = { loading: true };
    jest.spyOn(MoviesContext, 'useMoviesContext').mockImplementation(() => contextValues);
    const wrapper = shallow(<MovieList />);
    const h4 = wrapper.find('h4');
    expect(h4.text()).toEqual('...loading');
  });
test('Should display the no result text in case there are no results', () => {
    const contextValues = { loading: false, movies: [] };
    jest.spyOn(MoviesContext, 'useMoviesContext').mockImplementation(() => contextValues);
    const wrapper = shallow(<MovieList />);
    const h4 = wrapper.find('h4');
    expect(h4.text()).toEqual('Sorry, there are no results');
  });
test('Should display one movieItem component in case one movie is returned', () => {
    const contextValues = {
        loading: false,
        genres: {1: 'adventure'},
        movies: [{
            id: 1,
            title: "Niko's Movie",
            poster_path: "",
            release_date: "2019",
            overview: "Just a movie",
            vote_average: '10',
            genre_ids: [1],
    }] };
    jest.spyOn(MoviesContext, 'useMoviesContext').mockImplementation(() => contextValues);
    const wrapper = mount(<MovieList />);
    expect(wrapper.find('MovieItem')).toHaveLength(1);

  });
  test('Should match the snapshot', () => {
    const contextValues = {
      loading: false,
      genres: {1: 'adventure'},
      movies: [{
          id: 1,
          title: "Niko's Movie",
          poster_path: "",
          release_date: "2019",
          overview: "Just a movie",
          vote_average: '10',
          genre_ids: [1],
  }] };
  jest.spyOn(MoviesContext, 'useMoviesContext').mockImplementation(() => contextValues);
    const tree = renderer.create(<MovieList />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});