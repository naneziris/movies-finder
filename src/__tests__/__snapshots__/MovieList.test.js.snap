// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList Should match the snapshot 1`] = `
<div>
  <ul
    className="movies-container"
  >
    <div>
      <li
        className="movie "
      >
        <div
          className="movie__container"
        >
          <div
            className="movie__media"
          >
            <img
              alt="Niko's Movie"
              src="https://image.tmdb.org/t/p/w500"
            />
          </div>
          <div
            className="movie__body"
          >
            <h2
              className="movie__title"
            >
              Niko's Movie
            </h2>
            <h5
              className="movie__date"
            >
              2019
            </h5>
            <h5
              className="movie_average"
            >
              10
            </h5>
            <h3
              className="movie__description"
            >
              Description
              <p>
                Just a movie
              </p>
            </h3>
            <div>
              <span
                className="movies__genre"
              >
                adventure
              </span>
            </div>
            <button
              className="movie__btn"
              data-id={1}
              onClick={[Function]}
              type="button"
            >
              More Info
            </button>
          </div>
        </div>
      </li>
    </div>
  </ul>
</div>
`;
